### ISSUE TITLE
Tokenizer training sticks for long times.

### ISSUE THREAD

**Issue by lancelotzty** [huggingface/tokenizers#814]
When training a byelevelBPE for a text file of size 1.8GB, it takes about 5 hours for 90% of the count pairs, but hung for 2 hours. The process keeps running, but not all cpus are used. 
The code for training the tokeniser is 

```python
tokennizer = ByteLevelBPETokenizer(lowercase=True)
tokeniser.train(text_path,
                         min_frequencey=2,
                         show_progress=True,
                         special_tokens={"<s>", "<pad>", "</s>", "<unk>", "mask"}
                        )
```
<img width="1439" alt="error shot" src="https://user-images.githubusercontent.com/32071296/136530629-2660f0f2-4cdb-4c77-8211-75e2b8481999.png">
Can anyone help me on pointing out what might cause this issue? Thanks!

**Comment by dszhengyu** [comment]
Same here, I'm using AWS x1e.32xlarge (3,904 GB memory), training SentencePieceBPE on 180GB corpus, it just hang for ~8 hours on `Count pairs` on the last 1% (the progress bar froze completely):
```
[08:42:09] Pre-processing files (194139 Mo)         ��������������������������������������������������������������������������������                100%
[01:11:12] Tokenize words                           �������������������������������������������������������������������������������� 700608638/700608638
[05:27:56] Count pairs                              �������������������������������������������������������������������������������� 693602514/700608638
```
htop result:
<img width="1527" alt="Screen Shot 2022-04-16 at 14 25 22" src="https://user-images.githubusercontent.com/7588405/163664413-d1bad645-02c0-40f3-8fb5-6b30f7fbbf4d.png">

**Comment by Narsil** [comment]
Hi @dszhengyu ,

Thanks for the report.

- 180Go is likely to trigger some bug where we overflow the `u32` count method ( I can't be certain it will trigger, just really look at the result tokenization as if something overflows it might be silent and just ruin your tokenizer). There's no easy fix for that, as making the lib purely  `u64` is going to slow it down for many people and supporting both is a big change to do.
- Thanks a lot for the `top` screen. It's using only  1 core instead of 64, which could be the reason of this "stuck" behavior.
- It doesn't seem to be swap related.


The first issue needs to be addressed in a separate PR if you want to get a good tokenizer (no overflow), then we can a look at the slow step you're seeing, Then we can parallelize the loop for the heap see if that's the issue.

Is your data English ? Or a space separated langage at least (not Chinese for instance). If your pre_tokenization is not very efficient then there might be other issues related to BPE itself which doesn't really handle long sequences that well (without spaces, sequences can be quite long quite fast).

**Comment by saberiato** [comment]
Same issue here... it (`BpeTrainer`) stuck on "Compute merges", while consuming all 64 threads for a long time...

<img width="1401" alt="image" src="https://github.com/huggingface/tokenizers/assets/44996387/ffd69be7-1b95-412c-a02b-e552ff0a7378">
<img width="1850" alt="image" src="https://github.com/huggingface/tokenizers/assets/44996387/bf3e486a-009a-4376-b34a-33adc3675569">

**Comment by frotaur** [comment]
Same problem, stuck on compute merges for 10+hrs when all the other steps took ~ 3hrs. Dataset size if 300Gb, English. CPU is at 1000%

**Comment by EndlessCG** [comment]
Is there any update on this issue? Seems like BPE tokenizers get stuck at `Computing merges` on long training corpus.

**Comment by ArthurZucker** [comment]
#1313 and https://github.com/huggingface/tokenizers/issues/1206#issuecomment-1496022146 should give some ideas.

**Comment by MahdiMasoon** [comment]
I also have the issue of slow training speed with the tokenizer on smaller datasets. Upon investigation, it became clear that the tokenizer only utilizes 1 CPU core, and batching or not batching doesn’t affect its speed. What do you think is the solution to this problem?

**Comment by ArthurZucker** [comment]
Could you share a reproducer? If you have `os.environ["TOKENIZERS_PARALLELISM"] = "1" ` ( in python) or if you have the correct ray package for rust / depends on the tokenizer you are using as well and the version you are using

**Comment by MahdiMasoon** [comment]
> Could you share a reproducer? If you have `os.environ["TOKENIZERS_PARALLELISM"] = "1" ` ( in python) or if you have the correct ray package for rust / depends on the tokenizer you are using as well and the version you are using

Thank you for your guidance, it’s fixed!

**Comment by tehranixyz** [comment]
> > Could you share a reproducer? If you have `os.environ["TOKENIZERS_PARALLELISM"] = "1" ` ( in python) or if you have the correct ray package for rust / depends on the tokenizer you are using as well and the version you are using
> 
> Thank you for your guidance, it’s fixed!

Do you mind sharing your solution? How were you able to fix the issue? Thanks.

**Comment by sinaahmadi** [comment]
Checking on this issue, can someone share a workaround or insight into this problem?

**Comment by SoshyHayami** [comment]
just run
os.environ["TOKENIZERS_PARALLELISM"] = "1"  . it's weird but the BPE trainer has issues with parallelism.

**Comment by MahdiMasoon** [comment]
> just run os.environ["TOKENIZERS_PARALLELISM"] = "1" . it's weird but the BPE trainer has issues with parallelism.

@tehranixyz @sinaahmadi, I apologize for the delay in responding. As @SoshyHayami mentioned, executing this command will resolve the issue, and this process will run in parallel.

**Comment by kdcyberdude** [comment]
Hi guys, Any update on this... it's still damn slow. 

I am training a BPE tokenizer on 7.6M docs of approx 30-40GB of data. given the speed of `compute merges` it's gonna take 157 days to be completed🤐.

I have tried with and without TOKENIZERS_PARALLELISM.
```python

import os
os.environ["TOKENIZERS_PARALLELISM"] = "1"
from datasets import load_from_disk,load_dataset, concatenate_datasets
from transformers import AutoTokenizer
from tqdm import tqdm

ds = load_dataset('kdcyberdude/Vichaar', num_proc=4, cache_dir = './gemma_data_cache')['train']
tokenizer = AutoTokenizer.from_pretrained("openchat/openchat-3.5-0106-gemma")
print(len(tokenizer))
print(tokenizer.is_fast)

def batch_iterator(batch_size=1000):
    for i in range(0, 7600000, batch_size):
         yield ds[i : i + batch_size]["text"]

new_tokenizer = tokenizer.train_new_from_iterator(
    batch_iterator(),
    vocab_size=32_000,
)

new_tokenizer.save_pretrained("./gemma-32k-pa-tokenizer", )

```
![Screenshot from 2024-07-04 12-26-07](https://github.com/huggingface/tokenizers/assets/34835322/446a2c3e-acab-438a-b69f-9eb37fe011b9)

**Comment by ArthurZucker** [comment]
Hey! I'll see what I can do

**Comment by kdcyberdude** [comment]
@ArthurZucker, this is only happening in case of gemma, llama3 and T5 is working fine.
